lib/cognito-idp-stack.ts

```diff
@@ -0,0 +1,380 @@
+import * as cdk from '@aws-cdk/core';
+import * as lambda from '@aws-cdk/aws-lambda';
+import * as apigw from '@aws-cdk/aws-apigateway';
+import * as iam from '@aws-cdk/aws-iam';
+import * as cognito from '@aws-cdk/aws-cognito';
+import * as acm from '@aws-cdk/aws-certificatemanager';
+import { StaticSite } from './static-site';
+import { ResourceHandlerProps } from './resource-handler-props';
+import * as dynamodb from '@aws-cdk/aws-dynamodb';
+import * as secrets from '@aws-cdk/aws-secretsmanager';
+import * as route53 from '@aws-cdk/aws-route53';
+import * as targets from '@aws-cdk/aws-route53-targets';
+import { CognitoRestApiProps, CognitoRestApi } from './cognito-rest-api';
+import * as cr from '@aws-cdk/custom-resources';
+
+/**
+ * Environment variables needed to deploy the stack.
+ */
+export interface CognitoIdpStackProps extends cdk.StackProps {
+
+    /**
+     * Domain name for the web site, e.g. www.example.com
+     */
+    readonly webDomainName: string;
+
+    /**
+     * The ARN to the certificate for the web site.
+     */
+    readonly webCertificateArn: string;
+
+    /**
+     * The ARN to the FaceBook app secret.
+     */
+    readonly facebookSecretArn: string;
+
+    /**
+     * The FaceBook app id.
+     */
+    readonly facebookAppId: string;
+
+    /**
+     * The domain name for the REST API. e.g. api.example.com
+     * 
+     * This cannot include any reserved strings like 'aws' or 'cognito'.
+     */
+    readonly apiDomainName: string;
+
+    /**
+     * The ARN to the certificate for the REST API domain.
+     */
+    readonly apiCertificateArn: string;
+
+    /**
+     * The FaceBook API version.
+     */
+    readonly facebookApiVersion: string;
+
+    /**
+     * The redirect Uri for the Cognito auth code.
+     */
+    readonly cognitoRedirectUri: string;
+
+    /**
+     * The ID of the hosted zone for the web site domain.
+     */
+    readonly hostedZoneId: string;
+
+    /**
+     * The ID of the hosted zone for the API domain.
+     */
+    readonly apiHostedZoneId: string;
+
+    // Note: The below variables are only required for integration testing
+    // after deploying the stack, so you can wait to fill those in
+
+    readonly cognitoPoolId: string;
+    readonly cognitoDomainPrefix: string;
+    readonly cognitoAppClientId: string;
+    readonly cognitoRegion: string;
+    readonly userTable: string
+    readonly jwt: string;
+}
+
+/**
+ * This stack contains the following:
+ * 
+ * - The static site (S3, CloudFront, Route53)
+ * - The REST API (Lambda and API Gateway)
+ * - Auth (Cognito)
+ */
+export class CognitoIdpStack extends cdk.Stack {
```

Comment:
A recommended practice with the CDK is to model a stack into its constituent constructs. I like the `CognitoRestApi` and the `StaticStite` L3s. Would be nice to do more of such organizing to this large file.

They don't all have to directly fit the L3 reusable pattern but simply constructs that are customized for this app, and help keep the code organized.

For instance the user pool, domain, client, idp, etc. can become a construct, the data layer, i.e. DDB can be another one, CognitoRestApi, the custom domain and the lambda functions can be another.

Suggestion for better code organization 😊